### DISCUSSION TITLE
Why does openssl seemingly trim leading 0s in EC params?

### DISCUSSION THREAD

**Discussion by maszlasz** [openssl/openssl#24558] [Q&A] [answered]
e.g. when running 
`openssl ecparam -name secp521r1 -param_enc explicit -text -noout`
you get
```
EC-Parameters: (521 bit)
Field Type: prime-field
Prime:
    01:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:
    ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:
    ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:
    ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:
    ff:ff:ff:ff:ff:ff
A:   
    01:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:
    ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:
    ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:
    ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:ff:
    ff:ff:ff:ff:ff:fc
B:   
    51:95:3e:b9:61:8e:1c:9a:1f:92:9a:21:a0:b6:85:
    40:ee:a2:da:72:5b:99:b3:15:f3:b8:b4:89:91:8e:
    f1:09:e1:56:19:39:51:ec:7e:93:7b:16:52:c0:bd:
    3b:b1:bf:07:35:73:df:88:3d:2c:34:f1:ef:45:1f:
    d4:6b:50:3f:00
Generator (uncompressed):[...]
```
with the B parameter being one byte shorter, because its leading 00 has been trimmed.
Openssl 3.3.0 and earlier.

**Comment by levitte** [comment]
All three parameters (Prime, A, B) are (BIG) numbers, and are printed as such.  (BIG) numbers, just like any number, never conserve leading zeroes.

**Reply by levitte** [reply]
This isn't the *entire* truth, BTW.  *If* the highest byte in the number has the high bit set (in hex, that's `8` to `f`), then a leading `00` will be printed, to ensure it looks like a positive number.

**Reply by romen** [reply, marked as answer]
I was also answering, but @levitte was too fast.
If you want to dive more on why/how/when this happens, it has to do with the rules for DER encoding of (big) integers, which is the low level format standardized for representing the EC parameters.

**Reply by maszlasz** [reply]
Makes sense, thanks both of you!
